README.md
# DocManager
[![Build Status](https://travis-ci.org/openSUSE/docmanager.svg?branch=master)](https://travis-ci.org/openSUSE/docmanager)
[![codecov.io](http://codecov.io/github/openSUSE/docmanager/coverage.svg?branch=develop)](http://codecov.io/github/openSUSE/docmanager?branch=develop)
[![Code Health](https://landscape.io/github/openSUSE/docmanager/develop/landscape.svg?style=flat)](https://landscape.io/github/openSUSE/docmanager/develop)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/openSUSE/docmanager/badges/quality-score.png?b=develop)](https://scrutinizer-ci.com/g/openSUSE/docmanager/?branch=develop)

DocManager is a tool to manage meta-information within DocBook 5
documents. You can add, delete, edit, and query meta information.
Each meta-information is stored in the `info` element in the
DocManager namespace.

### Repositories
  1. Stable Repository: https://build.opensuse.org/package/show/Documentation:Tools/docmanager
  2. Unstable Repository: https://build.opensuse.org/package/show/Documentation:Tools:Develop/docmanager

### Usage
  1. **Adding or editing meta-information**

      ```
      $ docmanager set -p maintainer=test_name -p version=test_version test_file1.xml test_file2.xml
      ```

  2. **Deleting meta-information**

      ```
      $ docmanager del -p maintainer,version test_file1.xml test_file2.xml
      ```

  3. **Get meta-information**

      ```
      $ docmanager get -p maintainer test_file1.xml
      value_of_property_maintainer
      ```

  Do not provide properties if you want to get all information.

  4. **Analyze meta-information**

      ```
      $ docmanager analyze -qf "This file ({os.file}) is maintained by {maintainer} and has the priority {priority}." example1.xml example2.xml
      ```
      Example output for each file:
      ```
      This file (example1.xml) is maintained by mschnitzer and has the priority 10.
      This file (example2.xml) is maintained by toms and has the priority 5.
      ```

### Developers
- [Rick Salevsky](https://github.com/RSalevsky)
- [Manuel Schnitzer](https://github.com/mschnitzer)
- [Thomas Schraitle](https://github.com/tomschr)

### Contribution
  1. Create a branch if you have access otherwise fork it.
  2. Make your changes.
  3. Create a pull request.
  4. Done! Your changes will be reviewed as soon as possible.


ChangeLog
-------------------------------------------------------------------
Mon Aug 31 17:51:39 UTC 2015 - mschnitzer@suse.com

DocManager 3.3.3 Release

## Features
* Implemented 3 methods for handling XML attributes: set-attr, get-attr, and del-attr
* Added {os.lastmodify} to `analyze` feature

## Bugfixes
* `get` will now show sub properties if all properties are desired.

## Miscellaneous
* Many changes in man page 

-------------------------------------------------------------------
Tue Aug 25 13:26:21 UTC 2015 - mschnitzer@suse.com

## Bugfixes
* There was a logic error in our file list parser.

-------------------------------------------------------------------
Tue Aug 25 09:30:20 UTC 2015 - mschnitzer@suse.com

## Bugfixes
* DocManager hangs during the initialization of broken XML files #96
* Error output messages without file names #97

-------------------------------------------------------------------
Mon Aug 24 09:27:07 UTC 2015 - mschnitzer@suse.com

## Bugfixes
* Issues in test cases
* Help text for --default-output
* Default output for 'analyze' cmd
* Typo in config file (querformat -> queryformat)

## New features
* --stop-on-error and -q for 'analyze' cmd
* Runtime information
* Threading feature (--jobs/-j)- 

-------------------------------------------------------------------
Mon Aug 10 12:53:55 UTC 2015 - mschnitzer@suse.com

## Bugfixes
* Fix alias bug: It was not possible to add extra arguments to an alias
* --version shows now the version without any warnings- 

-------------------------------------------------------------------
Mon Aug 10 10:25:26 UTC 2015 - mschnitzer@suse.com

## Bugfixes
* There was a not caught exception in XmlHandler
* The output of `set` was disabled when no valid XML files were
  found. 

-------------------------------------------------------------------
Fri Aug  7 14:21:54 UTC 2015 - mschnitzer@suse.com

Stable release DocManager 3.2.0:

## Bugfixes
* Del Subcommand Should Really Delete Properties (#60)
* Output of not available properties is 'None' (#65)
* Resolve * in aliases (#67)
* Catch KeyboardInterrupt exception (#77)
* 'config' subcommand does not show 'Permission denied' error (#80)
* alias subcommand: "Permission denied for file '/' [...]" (#86)

## New features
* Adapt DocManager with Configuration File (#27)
* Implement Query/Analyze Feature (#49)
* Support Aliases (#64)
* Default aliases (#66)
* Implement 'config' subcommand (#70)
* Implement 'alias' subcommand (#72)
* Allow to set verbosity level through config file (#81)

## Miscellaneous
* documented several new features in manpage (#36, #82)
* Prepared setup.py for OBS (5f6990b)

-------------------------------------------------------------------
Thu Jul 23 15:43:28 UTC 2015 - toms@opensuse.org

Stable release DocManager 3.1.0:

* Bugfixes:
  - Fixed #58 (Ignore Directories in File List)
  - Fixed #57 (Ignoring XML Files in Case Root Element is Wrong)
  - Fixed #56 (Add Tracker Meta Information)
  - docmanager/xmlutil.py in findprolog():
    Set maxsize to -1 by default to avoid XML errors when extensive
    XML comments before root start tag are visible

* New features:
  - Issue #56: Add Tracker Meta Information
  - Extended option `--stop-on-error` for `init` subcommand

* Miscellaneous:
  - Updated man page (69c8037, 1b67884, ae221a6)

-------------------------------------------------------------------
Thu Jun 18 10:56:30 UTC 2015 - mschnitzer@suse.com

This is a stable release of DocManager.

ChangeLog:
* Performance improvement
* --repository is a new option in init and set (have a look into the man page)
* Updated man page and restructured some things
* Internal code improvement
* Fixed bug in del sub command
* Fixed some typos
* Adjusted the output of set and get
* Added table output format
* Added xml output format
* Added --stop-on-error as option (documentation is in the man page)
* init does now support the same options like in set (also documented in the man page)- 

-------------------------------------------------------------------
Tue Jun 16 10:45:11 UTC 2015 - mschnitzer@suse.com

Added 2 new requirements: python-lxml and python3-lxml 

-------------------------------------------------------------------
Tue Jun 16 10:42:59 UTC 2015 - mschnitzer@suse.com

Release of DocManager 3.0.0-RC4

* Code refactoring
* Added more valid root elements (chapter, appendix, and preface)
* Improved output message of "invalid root element"
* Performance improvement

docmanager/cli.py
#
# Copyright (c) 2015 SUSE Linux GmbH
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 3 of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, contact SUSE LLC.
#
# To contact SUSE about this file by physical or electronic mail,
# you may find current contact information at www.suse.com

import argparse
from glob import glob
import logging
import re
import shlex
import sys
import os
import os.path
import urllib.request
from configparser import NoOptionError, NoSectionError
from docmanager import __version__
from docmanager.config import docmanagerconfig, create_userconfig
from docmanager.core import ReturnCodes, LANGUAGES, STATUSFLAGS, \
    DefaultDocManagerProperties, BugtrackerElementList, DefaultSubCommands
from docmanager.exceptions import DMConfigFileNotFound
from docmanager.logmanager import log, logmgr_flog, setloglevel


def populate_properties(args):
    """Populate args.properties from "standard" options

    :param argparse.Namespace args:
    :return: list of property=value items
    :rtype: list
    """

    result = []
    for prop in DefaultDocManagerProperties:
        proparg = prop.replace("/", "_")
        if hasattr(args, proparg) and getattr(args, proparg) is not None:
            result.append("{}={}".format(prop, getattr(args, proparg)))

    return result


def populate_bugtracker_properties(args):
    """Populate args.bugtracker_* from "standard" options

    :param argparse.Namespace args:
    :return: list of property=value items
    :rtype: list
    """

    result = []
    for prop in BugtrackerElementList:
        proparg = prop.replace("/", "_")
        if hasattr(args, proparg) and getattr(args, proparg) is not None:
            result.append("{}={}".format(prop, getattr(args, proparg)))

    return result


def init_subcmd(subparsers, stop_on_error, propargs, mainprops, filesargs):
    """Create the 'init' subcommand

    :param subparsers:           Subparser for all subcommands
    :param dict stop_on_error:   Dict for the --stop-on-error option
    :param dict propargs:        Dict with action and help for default
                                 properties
    :param tuple mainprops:      Tuple of short and long options of default
                                 properties
    :param dict filesargs:       Dict for FILE argument
    """
    # 'init' command for the initialization
    pinit = subparsers.add_parser('init',
                                  aliases=['i'],
                                  help='Initializes an XML document with '
                                       'predefined properties.')
    pinit.add_argument('--force',
                       action='store_true',
                       help='This option forces the initialization.'
                       )
    pinit.add_argument('--stop-on-error', **stop_on_error)
    pinit.add_argument('--with-bugtracker',
                       action='store_true',
                       help='Adds a bugtracker structure to an XML file.'
                       )
    pinit.add_argument('-p', '--properties', **propargs)

    for options in mainprops:
        pinit.add_argument(*options,
                           help='Sets the property "{}"'.format(options[1][2:])
                           )

    # TODO: Do we really need that?
    pinit.add_argument('--repository',
                       help='Sets the property "repository".'
                       )
    for item in BugtrackerElementList:
        _, option = item.split('/')
        pinit.add_argument('--bugtracker-{}'.format(option),
                           help='Sets the property '
                                '"bugtracker/{}".'.format(option)
                           )

    pinit.add_argument("files", **filesargs)


def get_subcmd(subparsers, propargs, filesargs):
    """Create the 'get' subcommand

    :param subparsers:      Subparser for all subcommands
    :param dict propargs:   Dict with action and help for default
                            properties
    :param dict filesargs:  Dict for FILE argument
    """
    pget = subparsers.add_parser('get',
                                 aliases=['g'],
                                 help='Get key and returns value'
                                 )
    pget.add_argument('-p', '--properties', **propargs)
    pget.add_argument('-f', '--format',
                      choices=['table', 'json', 'xml'],
                      help='Set the output format.'
                      )
    pget.add_argument("files", **filesargs)


def set_subcmd(subparsers, stop_on_error, propargs, mainprops, filesargs):
    """Create the 'set' subcommand

    :param subparsers:           Subparser for all subcommands
    :param dict stop_on_error:   Dict for the --stop-on-error option
    :param dict propargs:        Dict with action and help for default
                                 properties
    :param tuple mainprops:      Tuple of short and long options of
                                 default properties
    :param dict filesargs:       Dict for FILE argument

    """
    pset = subparsers.add_parser('set',
                                 aliases=['s'],
                                 help='Set key=value property (one or more) '
                                      ' to delete the key let the value blank.'
                                 )
    pset.add_argument('-B', '--bugtracker',
                      action='store_true')
    pset.add_argument('--stop-on-error', **stop_on_error)
    pset.add_argument('-p', '--properties', **propargs)

    for options in mainprops:
        pset.add_argument(*options,
                          help='Sets the property "{}"'.format(options[1][2:])
                          )

    # TODO: Do we really need that?
    pset.add_argument('--repository',
                      help='Sets the property "repository"'
                      )

    for item in BugtrackerElementList:
        _, option = item.split('/')
        pset.add_argument('--bugtracker-{}'.format(option),
                          help='Set the property "bugtracker/{}" '
                               'for the given documents.'.format(option)
                          )

    pset.add_argument("files", **filesargs)


def del_subcmd(subparsers, propargs, filesargs):
    """Create the 'del' subcommand

    :param subparsers:      Subparser for all subcommands
    :param dict propargs:   Dict with action and help for
                            default properties
    :param dict filesargs:  Dict for FILE argument

    """
    pdel = subparsers.add_parser('del',
                                 aliases=['d'],
                                 help='Delete properties from XML documents'
                                 )
    pdel.add_argument('-p', '--properties', **propargs)
    pdel.add_argument("files", **filesargs)

def analyze_subcmd(subparsers, queryformat, filters, sort, default_output, filesargs):
    """Create the 'analyze' subcommand

    :param subparsers:           Subparser for all subcommands
    :param queryformat:          The queryformat
    :param dict filesargs:       Dict for FILE argument

    """
    panalyze = subparsers.add_parser('analyze',
                        aliases=['a'],
                        help='Analyzes the given XML files.'
                    )
    panalyze.add_argument('-qf', '--queryformat', **queryformat)
    panalyze.add_argument('-f', '--filter', **filters)
    panalyze.add_argument('-s', '--sort', **sort)
    panalyze.add_argument('-do', '--default-output', **sort)
    panalyze.add_argument("files", **filesargs)

def config_subcmd(subparsers):
    """Create the 'config' subcommand

    :param subparsers:           Subparser for all subcommands
    """

    pconfig = subparsers.add_parser('config', aliases=['c'], help='Modify tool for config files.')
    pconfig.add_argument('-s', '--system', action='store_true', help='Uses the system config file.')
    pconfig.add_argument('-u', '--user', action='store_true', help='Uses the user config file.')
    pconfig.add_argument('-r', '--repo', action='store_true', help='Uses the repository config file of the current repository.'
                                                                   ' (The user has to be in a git repository!)')
    pconfig.add_argument('-o', '--own', metavar='FILE', action='store', help='Uses a specified config file.')
    pconfig.add_argument('property', metavar='PROPERTY', help='Property (Syntax: secion.property)')
    pconfig.add_argument('value', metavar='VALUE', nargs='?', help='Value of the property.')

def alias_subcmd(subparsers):
    """Create the 'alias' subcommand

    :param subparsers:           Subparser for all subcommands
    """

    palias = subparsers.add_parser('alias', aliases=['al'], help='Modify tool for aliases.')
    palias.add_argument('alias_action', metavar='ACTION', choices=['set', 'del', 'get'], help='The action you want to perform.')
    cfgfile_group = palias.add_mutually_exclusive_group(required=True)
    cfgfile_group.add_argument('-s', '--system', action='store_const', const=1, dest="method", help='Uses the system config file.')
    cfgfile_group.add_argument('-u', '--user', action='store_const', const=2, dest="method", help='Uses the user config file.')
    cfgfile_group.add_argument('-r', '--repo', action='store_const', const=3, dest="method", help='Uses the repository config file of the current repository.'
                                                                   ' (The user has to be in a git repository!)')
    cfgfile_group.add_argument('-o', '--own', metavar='FILE', action='store', help='Uses a specified config file.')
    palias.add_argument('alias', metavar='ALIAS', help='Name of the alias')
    palias.add_argument('command', metavar='COMMAND', nargs='?', help='Command for the alias.')


def rewrite_alias(args):
    """Rewrite aliases

    :param argparse.Namespace args: Parsed arguments
    """
    actions = DefaultSubCommands
    args.action = actions.get(args.action)


def clean_filelist(args):
    """Clean the file list from unwanted directory names

    :param argparse.Namespace args: Parsed arguments
    """
    # Remove any directories from our files list
    allfiles = args.files[:]
    args.files = [f for f in args.files if not os.path.isdir(f)]
    diff = list(set(allfiles) - set(args.files))
    if diff:
        print("Ignoring the following directories:", ", ".join(diff))


def fix_properties(args):
    """Make different property styles consistent

    :param argparse.Namespace args: Parsed arguments
    """
    # Handle the different styles with -p foo and -p foo,bar
    # Needed to split the syntax 'a,b', 'a;b' or 'a b' into a list
    # regardless of the passed arguments
    _props = []
    # Use an empty list when args.properties = None
    args.properties = [] if args.properties is None else args.properties
    for item in args.properties:
        repl = re.split("[,;]", item)
        _props.extend(repl)
    args.properties = _props

    # Fill "standard" properties (like status) also into properties list:
    if args.action in ("s", "set"):
        args.properties.extend(populate_properties(args))
        args.properties.extend(populate_bugtracker_properties(args))

def is_alias(subcmd):
    """Check if subcmd is an alias
    :rtype: bool
    """
    return subcmd not in DefaultSubCommands

def parse_alias_value(value):
    """Parse pre defined constants

    :param str value: alias
    :return: parsed alias
    :rtype: str
    """
    return value.replace("{USER}", os.environ['USER'])

def fix_filelist(files):
    """Replaces * with all files in a directory (shell like)

    :param list files: file list from args.files
    """

    if files:
        for idx, i in enumerate(files):
            filelist = glob(i)
            if filelist:
                files.pop(idx)
                for x in filelist:
                    files.append(x)

def parsecli(cliargs=None, error_on_config=False):
    """Parse command line arguments

    :param list cliargs: Arguments to parse or None (=use sys.argv)
    :return: parsed arguments
    :rtype: argparse.Namespace
    """

    # parse just --config and --verbose
    confparser = argparse.ArgumentParser(add_help=False)
    confparser.add_argument('--config', dest='configfile', metavar='CONFIGFILE',
                            help='user config file location, uses also '
                                 'XDG_CONFIG_HOME env variable if set')
    confparser.add_argument('-v', '--verbose', action='count',
                            help="Increase verbosity level")
    args, remaining_argv = confparser.parse_known_args(cliargs)

    # Store configuration filename for further usage:
    configfile = args.configfile
    config = None

    # init log module
    setloglevel(args.verbose)

    # init config module
    # Exception handled in __init__.py
    config = docmanagerconfig(args.configfile)

    # Read the log level from the config files
    try:
        verbosity_level = int(config["general"]["verbosity_level"])
            
        if args.verbose is not None:
            if verbosity_level > args.verbose:
                args.verbose =  verbosity_level
        else:
            args.verbose = verbosity_level
        
        # set log level
        setloglevel(args.verbose)
    except KeyError:
        pass

    if remaining_argv:
        alias = remaining_argv[0]

        # parse aliases
        if is_alias(alias):
            try:
                value = parse_alias_value(config.get("alias", alias))
                cliargs = shlex.split(value)
            except (NoSectionError, NoOptionError) as err:
                log.warn(err)
                if error_on_config:
                    raise

    # parse cli parameters
    filesargs = dict(nargs='+',
                     metavar="FILE",
                     help='One or more DocBook XML or DC files.'
                     )
    propargs = dict(action='append',
                    # metavar="PROP[[=VALUE],PROP[=VALUE]...]
                    help='One or more properties to get, set, or delete. '
                         'Syntax of PROPERTIES: PROP[[=VALUE],PROP[=VALUE]...]'
                         ' Example (get/del): -p foo or -p foo,bar or '
                         '-p foo -p bar '
                         'Example (set): -p foo=a or -p foo=a,bar=b or '
                         '-p foo=a -p bar=b'
                    )
    stop_on_error = dict(action='store_true',
                       default=False,
                       help='Stop if an (XML) error is found '
                            'in a XML file (default: %(default)s)'
                )
    queryformat = dict(action='store',
                       help='The output query format. For more information, have a look into the manual page.'
                )
    sort = dict(action='store',
                       help='Sorts the output by XML properties.'
                )
    filters = dict(action='append',
                       help='Filters the analyzed data. For more information, have a look into the manual page.'
                )
    default_output = dict(action='store',
                       help='Sets the default output for properties which are not available in a file. By default, DocManager prints nothing.'
                )
    mainprops = tuple(("-{}".format(i.upper()[0]), "--{}".format(i))
                      for i in DefaultDocManagerProperties)

    parser = argparse.ArgumentParser(
        prog="docmanager",
        parents=[confparser],
        # usage="%(prog)s COMMAND [OPTIONS] FILE [FILE ...]\n",
        description="Docmanager sets, gets, deletes, or queries "
                    "meta-information for DocBook5 XML files.")
    parser.add_argument('--version',
                        action='version',
                        version='%(prog)s ' + __version__
                        )
    parser.add_argument('--langlist',
                        action='store_true'
                        )

    # Create a subparser for all of our subcommands,
    # save the subcommand in 'dest'
    subparsers = parser.add_subparsers(
        dest='action',
        title="Available subcommands",
        # metavar="COMMAND"
        )

    init_subcmd(subparsers, stop_on_error, propargs, mainprops, filesargs)
    get_subcmd(subparsers, propargs, filesargs)
    set_subcmd(subparsers, stop_on_error, propargs, mainprops, filesargs)
    del_subcmd(subparsers, propargs, filesargs)
    analyze_subcmd(subparsers, queryformat, filters, sort, default_output, filesargs)
    config_subcmd(subparsers)
    alias_subcmd(subparsers)

    # -----
    args = parser.parse_args(args=cliargs)

    if args.configfile is None:
        args.configfile = configfile

    args.config = config

    ##
    rewrite_alias(args)

    # Display language list
    if args.langlist is True:
        show_langlist()

    # If docmanager is called without anything, print the help and exit
    if args.action is None:
        parser.print_help()
        sys.exit(ReturnCodes.E_CALL_WITHOUT_PARAMS)

    if not hasattr(args, 'properties'):
        args.properties = None

    if not hasattr(args, 'stop_on_error'):
        args.stop_on_error = False

    # Clean file list
    if hasattr(args, 'files'):
        clean_filelist(args)
    else:
        args.files = None

    # Fix file list - this is needed for aliases - issue#67
    fix_filelist(args.files)

    # Fix properties
    fix_properties(args)

    # check for input format
    input_format_check(args)

    return args


def show_langlist(columns=None, padding=2):
    """Prints all supported languages

    :param int columns: Maximum number of characters in a column;
                        None to fill the current terminal window
    :param int padding: space from longest entry to border
    """

    try:
        from shutil import get_terminal_size
    except ImportError:
        import os

        def get_terminal_size(fallback=(80, 24)):
            return os.terminal_size(fallback)

    maxl = max([len(i) for i in LANGUAGES])

    if columns is None or columns < maxl:
        columns = get_terminal_size().columns
    length = len(LANGUAGES)
    rowwidth = maxl + padding + 1
    divisor = columns // rowwidth
    maxline = divisor * rowwidth

    fmt="".join(["{{:<{}}}|".format(maxl + padding) for _ in range(divisor)])# flake8: noqa
    line = "-"*maxline
    print(line)
    for start, stop in zip(range(0, length, divisor),
                           range(divisor, length+divisor, divisor)):
        x = list(LANGUAGES[start:stop])
        if len(x) < divisor:
            for i in range(divisor - len(x)):
                x.append("")
        print(fmt.format(*x))
    print(line)

    sys.exit(ReturnCodes.E_OK)


def input_format_check(args):
    """Checks if the given arguments have a correct value

    :param object args: Arguments object from argparser
    """
    logmgr_flog()

    if hasattr(args, 'status') and args.status is not None:
        if args.status not in STATUSFLAGS:
            print("Value of 'status' property is incorrect. "
                  "Expecting one of these values: "
                  "{}".format(", ".join(STATUSFLAGS)))
            sys.exit(ReturnCodes.E_WRONG_INPUT_FORMAT)

    if hasattr(args, 'deadline') and args.deadline is not None:
        r = re.match("^[0-9]{4}\-[0-9]{2}\-[0-9]{2}$", args.deadline)
        if r is None:
            print("Value of 'deadline' is incorrect. "
                  "Use this date format: YYYY-MM-DD")
            sys.exit(ReturnCodes.E_WRONG_INPUT_FORMAT)

    if hasattr(args, 'priority') and args.priority is not None:
        errmsg = ("Value of 'priority' is incorrect. "
                  "Expecting a value between 1 and 10.")

        if not args.priority.isnumeric():
            print(errmsg)
            sys.exit(ReturnCodes.E_WRONG_INPUT_FORMAT)

        args.priority = int(args.priority)
        if args.priority < 1 or args.priority > 10:
            print(errmsg)
            sys.exit(ReturnCodes.E_WRONG_INPUT_FORMAT)

    if hasattr(args, 'translation') and args.translation is not None:
        values = ('yes', 'no')
        if args.translation not in values:
            print("Value of 'translation' is incorrect. "
                  "Expecting one of these values: yes or no")
            sys.exit(ReturnCodes.E_WRONG_INPUT_FORMAT)

    if hasattr(args, 'languages') and args.languages is not None:
        for i in args.languages.split(","):
            if i not in LANGUAGES:
                print("Value of 'languages' is incorrect. "
                      "Language code '{}' is not supported. "
                      "Type 'docmanager --langlist' to see "
                      "all supported language codes.".format(i))
                sys.exit(ReturnCodes.E_WRONG_INPUT_FORMAT)

    if hasattr(args, 'repository') and args.repository is not None:
        request = None
        try:
            request = urllib.request.urlopen(args.repository)
        except ValueError:
            print("Value of 'repository' is incorrect. "
                  "The value is not a URL.")
            sys.exit(ReturnCodes.E_WRONG_INPUT_FORMAT)
        except urllib.error.URLError as err:
            if hasattr(err, 'code') and err.code is not None:
                if err.code is not 200:
                    log.warn("The remote server returns an error code "
                             "for this request: {} - Please double check if "
                             "the URL is correct. Nevertheless the URL will "
                             "be written into the "
                             "given files.".format(err.code))
            else:
                log.warn("The given URL '{}' seems to be invalid or the "
                         "remote server is not online. Please double "
                         "check if the URL is correct. Nevertheless the "
                         "URL will be written into the "
                         "given files.".format(args.repository))

        if hasattr(request, 'close'):
            request.close()
